File: lib/src/main/java/com/fanwe/lib/http/utils/TransmitParam.java
package com.fanwe.lib.http.utils;

/**
 * Created by zhengjun on 2017/10/13.
 */

public class TransmitParam
{
    private static final long CALCULATE_SPEED_INTERVAL = 50;

    private long mCurrent;
    private long mTotal;
    private int mProgress;
    private int mSpeedBps;

    private long mLastTime;
    private long mLastCount;

    private long mCalculateSpeedInterval = CALCULATE_SPEED_INTERVAL;

    public synchronized void transmit(long current, long total)
    {
        mCurrent = current;
        mTotal = total;

        final long currentTime = System.currentTimeMillis();
        final long timeInterval = currentTime - mLastTime;
        if (timeInterval >= mCalculateSpeedInterval)
        {
            long count = current - mLastCount;
            mSpeedBps = (int) (count * (1000f / timeInterval));

            mLastTime = currentTime;
            mLastCount = current;
        }

        mProgress = (int) (current * 100 / total);
    }

    public synchronized boolean isFinish()
    {
        return mCurrent == mTotal && mCurrent > 0;
    }

    public synchronized void setCalculateSpeedInterval(long calculateSpeedInterval)
    {
        if (calculateSpeedInterval <= 0)
        {
            calculateSpeedInterval = CALCULATE_SPEED_INTERVAL;
        }
        mCalculateSpeedInterval = calculateSpeedInterval;
    }

    public long getCurrent()
    {
        return mCurrent;
    }

    public long getTotal()
    {
        return mTotal;
    }

    public int getProgress()
    {
        return mProgress;
    }

    public int getSpeedBps()
    {
        return mSpeedBps;
    }

    public int getSpeedKBps()
    {
        return getSpeedBps() / 1024;
    }

    @Override
    public String toString()
    {
        StringBuilder sb = new StringBuilder();
        sb.append(getCurrent()).append("/").append(getTotal()).append("\r\n")
                .append(getProgress()).append("%").append("\r\n")
                .append(getSpeedKBps()).append("KBps");
        return sb.toString();
    }
}

File: lib/src/main/java/com/sd/lib/http/utils/TransmitParam.kt
package com.sd.lib.http.utils

class TransmitParam {
    private val mCalculateSpeedInterval: Long
    private var mLastTime: Long = 0
    private var mLastCount: Long = 0

    @JvmOverloads
    constructor(calculateSpeedInterval: Long = 0) {
        val interval = if (calculateSpeedInterval > 0) calculateSpeedInterval else 100
        mCalculateSpeedInterval = interval
    }

    /** 当前传输量 */
    var current: Long = 0
        private set

    /** 总量 */
    var total: Long = 0
        private set

    /** 传输进度 */
    var progress = 0
        private set

    /** 传输速率(Bps) */
    var speedBps = 0
        private set

    /** 传输速率(KBps) */
    val speedKBps: Int
        get() = speedBps / 1024

    /** 传输是否完成 */
    val isComplete: Boolean
        get() = current > 0 && current == total

    /**
     * 拷贝对象
     */
    fun copy(): TransmitParam {
        val copy = TransmitParam(mCalculateSpeedInterval)
        copy.current = current
        copy.total = total
        copy.progress = progress
        copy.speedBps = speedBps
        copy.mLastTime = mLastTime
        copy.mLastCount = mLastCount
        return copy
    }

    /**
     * 传输
     *
     * @param total 总量
     * @param current 当前传输量
     * @return true-进度发生了变化
     */
    @Synchronized
    fun transmit(total: Long, current: Long): Boolean {
        val oldProgress = progress
        if (total <= 0 || current <= 0) {
            reset()
            return oldProgress != progress
        }

        this.total = total
        this.current = current

        val currentTime = System.currentTimeMillis()
        val interval = currentTime - mLastTime
        if (interval >= mCalculateSpeedInterval) {
            val count = current - mLastCount
            speedBps = (count * (1000f / interval)).toInt()
            mLastTime = currentTime
            mLastCount = current
        }

        progress = (current * 100 / total).toInt()
        return progress > oldProgress
    }

    private fun reset() {
        current = 0
        total = 0
        progress = 0
        speedBps = 0
        mLastTime = 0
        mLastCount = 0
    }

    override fun toString(): String {
        return "${current}/${total} ${progress}% ${speedKBps}KBps ${super.toString()}"
    }
}